packages/flutter/lib/src/cupertino/segment_control.dart

```diff
@@ -0,0 +1,191 @@
+
+    for (int i = 0; i < widget.tabs.length; i++) {
+      SegmentControlItem tap = widget.tabs[i];
+      bool isActive = tap == widget.tabs[_activeTabIndex];
+      _ButtonPlace place = _ButtonPlace.start;
+
+      if (i > 0 && (widget.tabs.length - 1 == i)) {
+        place = _ButtonPlace.end;
+      } else if (i > 0 && (widget.tabs.length - 1 > i)) {
+        place = _ButtonPlace.middle;
+      }
+
+      list.add(new _SegmentControlItem(this, tap, place, isActive));
+    }
+
+    return new Column(
+      mainAxisAlignment: MainAxisAlignment.start,
+      children: <Widget>[
+        new Padding(
+          child: new Row(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: list,
+          ),
+          padding: new EdgeInsets.all(12.0),
+        ),
+        activeTab
+      ],
+    );
+  }
+
+  void changeTab(String title) {
+    setState(() {
+      for (int i = 0; i < widget.tabs.length; i++) {
+        SegmentControlItem t = widget.tabs[i];
+        if (t.title == title) {
+          _activeTabIndex = i;
+        }
+      }
+    });
+  }
+}
+
+class _SegmentControlItem extends StatefulWidget {
+  final double _defaultBorderRadius = 3.0;
+
+  final SegmentControlItem cupertinoButtonTab;
+  final _SegmentControlState parent;
+  final _ButtonPlace place;
+  final bool isActive;
+  final Color color;
+  final Color inverseColor;
+
+  _SegmentControlItem(
+      this.parent, this.cupertinoButtonTab, this.place, this.isActive,
+      {this.color = CupertinoColors.activeBlue,
+      this.inverseColor = CupertinoColors.white});
+
+  @override
+  State createState() {
+    return new _SegmentControlItemState(color, inverseColor);
+  }
+}
+
+class _SegmentControlItemState
+    extends State<_SegmentControlItem> {
+  Color color;
+  Color inverseColor;
+  bool tapDown = false;
+
+  _SegmentControlItemState(this.color, this.inverseColor);
+
+  BoxDecoration _boxDecoration(_ButtonPlace place) {
+    BorderRadius radius;
+
+    switch(place) {
+      case _ButtonPlace.start:
+        radius = new BorderRadius.only(
+          topLeft: new Radius.circular(widget._defaultBorderRadius),
+          bottomLeft: new Radius.circular(widget._defaultBorderRadius),
+        );
+        break;
+      case _ButtonPlace.end:
+        radius = new BorderRadius.only(
+          topRight: new Radius.circular(widget._defaultBorderRadius),
+          bottomRight: new Radius.circular(widget._defaultBorderRadius),
+        );
+        break;
+      default:
+        break;
+    }
+
+    BoxDecoration dec = new BoxDecoration(
+      color: widget.isActive ? color : inverseColor,
+      border: place == _ButtonPlace.middle
+          ? new Border(
+              top: new BorderSide(color: tapDown ? inverseColor : color),
+              bottom: new BorderSide(color: tapDown ? inverseColor : color),
+            )
+          : new Border.all(color: tapDown ? inverseColor : color),
+      borderRadius: radius,
```

Comment:
in practice you're probably going to want to implement this using a custom painter rather than a decoration.

Reply:
Why?

Reply:
because you need the vertical lines as well. You really want to draw the whole thing as one shape and draw it all at once so that the antialiasing and so forth ends up working correctly. If you use multiple decorations, they'll have weird artifacts where they meet up and it won't quite look right.

It'll also just be simpler code, probably. :-)